lib/wasmer-types/src/indexes.rs

```diff
@@ -25,60 +36,149 @@ entity_impl!(LocalMemoryIndex);
 #[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+#[cfg_attr(
+    feature = "enable-rkyv",
+    derive(RkyvSerialize, RkyvDeserialize, Archive)
+)]
+#[cfg_attr(
+    feature = "enable-rkyv",
+    archive(derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Debug))
+)]
 pub struct GlobalIndex(u32);
 entity_impl!(GlobalIndex);
+entity_impl!(ArchivedGlobalIndex);
 
 /// Index type of a linear memory (imported or local) inside the WebAssembly module.
 #[derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Debug)]
 #[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+#[cfg_attr(
+    feature = "enable-rkyv",
+    derive(RkyvSerialize, RkyvDeserialize, Archive)
+)]
+#[cfg_attr(
+    feature = "enable-rkyv",
+    archive(derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Debug))
+)]
 pub struct MemoryIndex(u32);
 entity_impl!(MemoryIndex);
+entity_impl!(ArchivedMemoryIndex);
 
 /// Index type of a signature (imported or local) inside the WebAssembly module.
 #[derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Debug)]
 #[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+#[cfg_attr(
+    feature = "enable-rkyv",
+    derive(RkyvSerialize, RkyvDeserialize, Archive)
+)]
+#[cfg_attr(
+    feature = "enable-rkyv",
+    archive(derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Debug))
+)]
 pub struct SignatureIndex(u32);
 entity_impl!(SignatureIndex);
+entity_impl!(ArchivedSignatureIndex);
 
 /// Index type of a passive data segment inside the WebAssembly module.
 #[derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Debug)]
 #[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+#[cfg_attr(
+    feature = "enable-rkyv",
+    derive(RkyvSerialize, RkyvDeserialize, Archive)
+)]
+#[cfg_attr(
+    feature = "enable-rkyv",
+    archive(derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Debug))
+)]
 pub struct DataIndex(u32);
 entity_impl!(DataIndex);
+entity_impl!(ArchivedDataIndex);
 
 /// Index type of a passive element segment inside the WebAssembly module.
 #[derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Debug)]
 #[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+#[cfg_attr(
+    feature = "enable-rkyv",
+    derive(RkyvSerialize, RkyvDeserialize, Archive)
+)]
+#[cfg_attr(
+    feature = "enable-rkyv",
+    archive(derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Debug))
+)]
 pub struct ElemIndex(u32);
 entity_impl!(ElemIndex);
+entity_impl!(ArchivedElemIndex);
 
 /// Index type of a custom section inside a WebAssembly module.
 #[derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Debug)]
 #[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+#[cfg_attr(
+    feature = "enable-rkyv",
+    derive(RkyvSerialize, RkyvDeserialize, Archive)
+)]
+#[cfg_attr(
+    feature = "enable-rkyv",
+    archive(derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Debug))
+)]
```

Comment:
I'm wondering at some point whether we should not add this inside the `entity_impl!` macro directly 🤔.

Reply:
It makes sense to add so. entity_impl! are all on some struct(integer), and they can trivially derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Debug)

Reply:
This was required because rkyv macro generate ArchivedCustomSectionIndex struct on CustomSectionIndex, then used ArchivedCustomSectionIndex and primary map key, but this key need Copy, ..., traits, so `archive(derive(Copy...))` is to let rkyv macro add a `derive(Copy,...)` on generated ArchivedCustomSectionIndex definition